File: fm/apps/alarm/plugins/subscribers.py
# -*- coding: utf-8 -*-
##----------------------------------------------------------------------
## Subscribers plugin
##----------------------------------------------------------------------
## Copyright (C) 2007-2016 The NOC Project
## See LICENSE for details
##----------------------------------------------------------------------

## NOC modules
from base import AlarmPlugin
from noc.lib.text import split_alnum
from noc.inv.models.interface import Interface


class SubscribersPlugin(AlarmPlugin):
    name = "subscribers"

    def get_data(self, alarm, config):
        # Get subscriber interfaces
        interfaces = []
        for i in Interface.objects.filter(
            managed_object=alarm.managed_object.id
        ):
            if i.profile.is_customer:
                interfaces += [i]
        if not interfaces:
            return {}  # No customers
        return {
            "plugins": [("NOC.fm.alarm.plugins.Subscribers", {})],
            "subscribers": [{
                "interface": i.name,
                "profile": i.profile.name,
                "description": i.description
            } for i in sorted(interfaces, key=split_alnum)]
        }

File: fm/apps/alarm/js/plugins/Subscribers.js
//---------------------------------------------------------------------
// Subscribers plugin
//---------------------------------------------------------------------
// Copyright (C) 2007-2016 The NOC Project
// See LICENSE for details
//---------------------------------------------------------------------
console.debug("Defining NOC.fm.alarm.plugins.Subscribers");

Ext.define("NOC.fm.alarm.plugins.Subscribers", {
    extend: "Ext.panel.Panel",
    title: "Subscribers",
    app: null,
    autoScroll: true,
    bodyPadding: 4,

    initComponent: function() {
        var me = this;

        me.store = Ext.create("Ext.data.Store", {
            fields: [
                "interface", "profile", "description"
            ],
            data: []
        });

        me.grid = Ext.create("Ext.grid.Panel", {
            autoScroll: true,
            store: me.store,
            columns: [
                {
                    dataIndex: "interface",
                    text: "Interface",
                    width: 150
                },
                {
                    dataIndex: "profile",
                    text: "Profile",
                    width: 150
                },
                {
                    dataIndex: "description",
                    text: "Description",
                    flex: 1
                }
            ]
        });

        Ext.apply(me, {
            items: [
                me.grid
            ]
        });

        me.callParent();
    },

    updateData: function(data) {
        var me = this;
        me.store.loadData(data.subscribers);
    }
});
